### DISCUSSION TITLE
[RESOLVED] Difference between getAnArgument and getArgument

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [github/codeql#7693] [General]
I'm wondering the difference between getAnArgument and getArgument? From the document there's a keyword "invocation", but since codeql does not execute the code so I'm not sure what it means here.

**Comment by intrigus-lgtm** [comment]
Have a look at the documentation:
https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/Expr.qll/predicate.Expr$Call$getAnArgument.0.html
https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/Expr.qll/predicate.Expr$Call$getArgument.1.html

Essentially:
`getArgument(_) = getAnArgument()` where `_` represents a [don't care](https://codeql.github.com/docs/ql-language-reference/expressions/#don-t-care-expressions) expression.

**Comment by smowton** [comment]
getAnArgument matches all arguments to the given call. Don't forget that all CodeQL expressions have the potential to be multi-valued: for example, if I had a program `f(1, "hello world");` and queried it with something like `from Call c where c.getCallee().getName() = "f" select c.getAnArgument()`, I would get a table of two results: `1` and `"hello world"`.

On the other hand `getArgument(index)` returns only arguments at a particular position. `select c.getArgument(1)` would return only `"hello world"`. `c.getArgument(index)` is expected to be single-valued if `c` and `index` both are.
